src/ILCompiler.ReadyToRun/src/IL/ReadyToRunILProvider.cs

```diff
@@ -9,19 +9,140 @@
+                MethodDesc compareExchangeObject = method.OwningType.GetKnownMethod("CompareExchange",
+                    new MethodSignature(
+                        MethodSignatureFlags.Static,
+                        genericParameterCount: 0,
+                        returnType: objectType,
+                        parameters: new TypeDesc[] { objectType.MakeByRefType(), objectType, objectType }));
+
+                ILEmitter emit = new ILEmitter();
+                ILCodeStream codeStream = emit.NewCodeStream();
+                codeStream.EmitLdArg(0);
+                codeStream.EmitLdArg(1);
+                codeStream.EmitLdArg(2);
+                codeStream.Emit(ILOpcode.call, emit.NewToken(compareExchangeObject));
+                codeStream.Emit(ILOpcode.ret);
+                return emit.Link(method);
+            }
+
+            return null;
+        }
+
+        private MethodIL TryGetIntrinsicMethodILForActivator(MethodDesc method)
+        {
+            if (method.Instantiation.Length == 1
+                && method.Signature.Length == 0
+                && method.Name == "CreateInstance")
+            {
+                TypeDesc type = method.Instantiation[0];
+                if (type.IsValueType && type.GetDefaultConstructor() == null)
+                {
+                    // Replace the body with implementation that just returns "default"
+                    MethodDesc createDefaultInstance = method.OwningType.GetKnownMethod("CreateDefaultInstance", method.GetTypicalMethodDefinition().Signature);
+                    return GetMethodIL(createDefaultInstance.MakeInstantiatedMethod(type));
+                }
+            }
+
+            return null;
+        }
+
+        private MethodIL TryGetIntrinsicMethodILForRuntimeHelpers(MethodDesc method)
+        {
+            if (method.Name == "IsReferenceOrContainsReferences")
+            {
+                TypeDesc type = method.Instantiation[0];
+                if (type.IsGCPointer)
+                    return new ILStubMethodIL(method, new byte[] { (byte)ILOpcode.ldc_i4_1, (byte)ILOpcode.ret }, Array.Empty<LocalVariableDefinition>(), Array.Empty<object>());
+                else
+                    return new ILStubMethodIL(method, new byte[] { (byte)ILOpcode.ldc_i4_0, (byte)ILOpcode.ret }, Array.Empty<LocalVariableDefinition>(), Array.Empty<object>());
+            }
+
+            // Ideally we could detect automatically whether a type is trivially equatable
+            // (i.e., its operator == could be implemented via memcmp). But for now we'll
+            // do the simple thing and hardcode the list of types we know fulfill this contract.
+            // n.b. This doesn't imply that the type's CompareTo method can be memcmp-implemented,
+            // as a method like CompareTo may need to take a type's signedness into account.
+            if (method.Name == "IsBitwiseEquatable")
```

Comment:
Should this be in the shared part?

Reply:
I am wondering whether it would make sense to make one and only IL intrinsics provider logic that is shared by all CoreLibs. Some of the intrinsics maybe unused for some flavors of the CoreLibs - that's ok.

Reply:
Well the problem is that these are different between CoreCLR and CoreRT. In CoreRT we have a usable method body for `IsBitwiseEquatable` and we only expand the intrinsic sometime. On CoreCLR the method body just throws so we have to always expand.

That's why I split these in the first place - there's subtle differences that make things confusing. E.g. the way CoreRT does comparers/equality comparers.

Reply:
> On CoreCLR the method body just throws so we have to always expand.

If we start doing that for CoreRT, it will work fine too and it may have better performance in some cases.